include/zephyr/usb/class/usbd_midi.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2024 Titouan Christophe
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_USB_CLASS_USBD_MIDI_H_
+#define ZEPHYR_INCLUDE_USB_CLASS_USBD_MIDI_H_
+
+/**
+ * @brief USB-MIDI 2.0 class device API
+ * @defgroup usb_midi USB MIDI 2.0 Class device API
+ * @ingroup usb
+ * @since 4.1
+ * @version 0.1.0
+ * @see midi20: "Universal Serial Bus Device Class Definition for MIDI Devices"
+ *              Document Release 2.0 (May 5, 2020)
+ * @{
+ */
+
+#include <zephyr/device.h>
+
+/**
+ * @brief      Send a Universal MIDI Packet to the host
+ * @param[in]  dev   The USB-MIDI interface
+ * @param[in]  ump   The packet to send, in Universal MIDI Packet format
+ * @return     0 on success
+ *             -EIO if MIDI2.0 is not enabled by the host
+ *             -EAGAIN if there isn't room in the transmission buffer
+ */
+int usbd_midi_send(const struct device *dev, const uint32_t ump[static 4]);
```

Comment:
Wow didn't even know the `[static 4]` thing was a thing. So I guess it's C99 (https://www.ibm.com/docs/en/i/7.5?topic=pd-static-array-indices-in-function-parameter-declarations-c-only), I see it's used in other spots in USB drivers which I suppose is fine but if C++ is not having it (in who knows what standards...) I reckon we should avoid it in headers (anything in `include/` really)

Maybe this could be a

```c
struct usbd_midi_ump {
        uint32_t data[4];
}
```
?

Reply:
I wanted to avoid the extra indirection (having to write `ump->data` everywhere), but I guess that would be the least worse compromise here.

Waiting for @jfischer-no re-review to take a decision on this

Reply:
Yeah ok I understand the intent now but then I really don't think it's a great idea: as designed the API has this logic where it extracts the data length from data packed in the data itself, this I think makes it harder to ensure that it's used correctly, and it exposes some potential out of bound read if data is not correct.

What I think this should do instead is to take a `uint8_t *buf` and `size_t len` as arguments, then the caller can trivially pass the buffer and `sizeof()` of such buffer, the function now instead of extracting the size from the data and assuming that the data is valid can extract the size from the data and compare it with the buffer length passed by the application and validate that the data buffer is big enough. That way the implementation becomes defensive to passing random data, I suppose worst case it would send garbage down the bus but at least it won't read out of buffer or crash as long as the caller does not lie about the buffer length.